packages/shylu/shylu_dd/bddc/src/ProblemMakerBDDC.h

```diff
@@ -149,86 +199,31 @@ template <class LO, class GO, class SX, class SM> class ProblemMaker
     return &m_nodeGIDs[0];
   }
 
-  const LO* getRowBegin() const {
-    return &m_rowBegin[0];
-  }
-
-  const LO* getColumns() const {
-    return &m_columns[0];
-  }
-
-  const SX* getValues() const {
-    return &m_values[0];
-  }
-
   const SX* getLoadVector() const {
     return &m_loadVector[0];
   }
 
-  void printElementMatrices() const
+  void printElementMatrices(const int elem=0) const
   {
-    printElementMatrix(&m_elemStiffMatrix[0], "stiffMatrix.dat");
-    printElementMatrix(&m_elemMassMatrix[0], "massMatrix.dat");
+    std::vector<SM> elemStiffMatrix, elemMassMatrix;
+    const double E(1.0), nu(0.3);
+    getElementMatrices(m_elemConn[elem], E, nu, elemStiffMatrix, elemMassMatrix);
+    printElementMatrix(elemStiffMatrix.data(), "stiffMatrix.dat");
+    printElementMatrix(elemMassMatrix.data(), "massMatrix.dat");
   }
 
-  void printSubdomainMatrix(const char* fnameBase,
-			    int ID) const
+  void getElementMatrices(std::vector<SM> & elemStiffMatrix,
+			  std::vector<SM> & elemMassMatrix,
+			  const int elem=0)
   {
-    char fname[101];
-    sprintf(fname, "%s%d.dat", fnameBase, ID);
-    LO i, j;
-    std::ofstream fout;
-    fout.open(fname);
-    LO numDof = m_nodeBegin[m_numNode];
-    for (LO i=0; i<numDof; i++) {
-      for (LO j=m_rowBegin[i]; j<m_rowBegin[i+1]; j++) {
-	fout << i+1 << " ";
-	fout << m_columns[j]+1 << " ";
-	SM realPart, imagPart;
-	getParts(m_values[j], realPart, imagPart);
-	fout << std::setw(22) << std::setprecision(15);
-	fout << realPart;
-	if (fabs(imagPart) != 0) {
-	  fout << " " << imagPart;
-	}
-	fout << std::endl;
-      }
-    }
-    fout.close();
+    const double E(1.0), nu(0.3);
+    getElementMatrices(m_elemConn[elem], E, nu, elemStiffMatrix,
+		       elemMassMatrix);
   }
 
-  void getConstrainEquations(LO & numRowsCT,
-			     LO & numColsCT,
-			     const LO* rowsCT,
-			     const LO* colsCT,
-			     const LO* rowBeginCT,
-			     const LO* columnsCT,
-			     const SM* matrixCT) const
-  {
-    numRowsCT = m_rowsCT.size();
-    numColsCT = m_colsCT.size();
-    rowsCT = &m_rowsCT[0];
-    colsCT = &m_colsCT[0];
-    rowBeginCT = &m_rowBeginCT[0];
-    columnsCT = &m_columnsCT[0];
-    matrixCT = &m_matrixCT[0];
-  }
-
-  void matrixVectorProduct(const SX* X, 
-			   int numVectors, 
-			   SX* AX) const
+  const std::vector< std::vector<LO> > getElementConnectivity() const
```

Comment:
Would it be a lot of overhead to return a new copy of these data each time?

Reply:
I was at first confused about this one, but now see that returning a reference could be better. This only gets called once early on during problem setup, but I may just go ahead and change this in the next go around.